EFCache.Redis/RedisCache.cs

```diff
@@ -1,38 +1,152 @@
+		table.insert(queryKeys, queryKey)
+		redis.call('del', queryKey)
+		redis.call('srem', entitySetKey, queryKey)
+	 end
+end
+return queryKeys";
+		private static byte[] _invalidateSetsScriptSha1;
+
+		// Script to delete a single query entry and to remove it from any entity sets in which
+		// it is contained.
+		// Accepts the query key as the first key arg and the entity set keys as the remaining key args
+		private const string InvalidateSingleKeyScript =
+@"local queryKey = KEYS[1]
+for i = 2,table.getn(KEYS) do
+	 redis.call('srem', KEYS[i], queryKey)
+end
+redis.call('del', queryKey)";
+		private static byte[] _invalidateSingleyKeyScriptSha1;
+
+		// Script to create a new query/result set entry. Add the string entry itself and add
+		// entries to any result sets used by the query
+		// Accepts the query key as the first key arg and the entity set keys as the remaining key args
+		private const string PutItemScript =
+@"local rsKey = KEYS[1]
+for i = 2,table.getn(KEYS) do
+	 redis.call('sadd', KEYS[i], rsKey)
+end
+redis.call('set', rsKey, ARGV[1])";
+		private static byte[] _putItemScriptSha1;
+
+		#endregion
+
+		#region Static Initialization
+
+		/// <summary>
+		/// Get the connection from the Lazy initializer
+		/// </summary>
+		private static ConnectionMultiplexer Connection => LazyRedis.Value;
 
-		private readonly string _cacheIdentifier;
-		private readonly string _statsIdentifier;
+		/// <summary>
+		/// Depending on how the instance was constructed, create a new (or use the existing)
+		/// connection multiplexer. Load the Lua scripts and save the hashes for calling before
+		/// returing the lazy initialized connection multiplexer
+		/// </summary>
+		/// <returns>A lazy initialized connection multiplexer that will load the Lua scripts on first use</returns>
+		private static Lazy<ConnectionMultiplexer> ConnectAndLoadScripts()
+		{
+			return _redis != null
+				? new Lazy<ConnectionMultiplexer>(() => LoadScripts(_redis))
+				: new Lazy<ConnectionMultiplexer>(() =>
+					LoadScripts(_redisConfig != null
+						? ConnectionMultiplexer.Connect(_redisConfig)
+						: ConnectionMultiplexer.Connect(_configurationOptions)));
+		}
 
-		private static ConnectionMultiplexer Connection => LazyRedis.Value;
+		/// <summary>
+		/// Load the Lua scripts into every instance of the Redis server in the cluster
+		/// </summary>
+		/// <param name="redis">The instance of the connection muliplexer</param>
+		/// <returns>The provided instance of the connection multiplexer</returns>
+		private static ConnectionMultiplexer LoadScripts(ConnectionMultiplexer redis)
+		{
+			var endpoints = redis.GetEndPoints();
+			foreach (var endpoint in endpoints)
+			{
+				var server = redis.GetServer(endpoint);
+				_invalidateSetsScriptSha1 = server.ScriptLoad(InvalidateSetsScript);
+				_invalidateSingleyKeyScriptSha1 = server.ScriptLoad(InvalidateSingleKeyScript);
+				_putItemScriptSha1 = server.ScriptLoad(PutItemScript);
```

Comment:
@jkdufair will overwriting these Sha1 values inside this loop cause a problem? what happens if there is more than one endpoint?

Reply:
It's the same script, so it'll just keep writing the same SHA1. I initially created my own hash, but I wanted to use whatever the server generates instead so there isn't weird line ending differences or whatever. So, no, it won't hurt. The loop is just if there are multiple master endpoints.

Reply:
@builtbywill ^